mlir/lib/Dialect/MIOpen/Transforms/ThreadwiseToStdlib.cpp

```diff
@@ -246,18 +246,51 @@ struct BufferLoadRewritePattern : public OpRewritePattern<BufferLoadOp> {
         oobDims.push_back(pair.index());
     bool toEmitOobChecks = !oobDims.empty();
 
-    auto emitLoadInstruction = [&b, loc, loadedType,
+    auto emitLoadInstruction = [&b, loc, loadedType, sourceType,
                                 source](ValueRange loadCoords) -> Value {
       Value loadedValue;
-      if (auto loadedVectorType = loadedType.dyn_cast<VectorType>()) {
-        // Issue vector load.
-        // use buffer load since the source memref is on address space 0
-        SmallVector<Value, 4> srcLowerIndicesI32;
-        for (auto v : loadCoords)
-          srcLowerIndicesI32.push_back(
-              b.create<IndexCastOp>(loc, b.getIntegerType(32), v));
-        loadedValue = b.create<gpu::MubufLoadOp>(loc, loadedType, source,
-                                                 srcLowerIndicesI32);
+
+      Type elementType = loadedType;
+      int64_t vectorLength = 1;
+
+      if (loadedType.isa<VectorType>()) {
+        VectorType loadedVectorType = loadedType.cast<VectorType>();
+        elementType = loadedVectorType.getElementType();
+        vectorLength = loadedVectorType.getShape()[0];
+
+        auto loadWidth = vectorLength * elementType.getIntOrFloatBitWidth();
+        bool isTooNarrow = loadWidth < 32;
+        bool isTooWide = loadWidth > 4 * 32;
+
+        // Use scalar load when load width is too narrow or too wide
+        // for a mubuf instruction
```

Comment:
Would it be better to hide this logic in the lowering logic for `gpu.mubuf_load`?

Reply:
I'm assuming `MubufLoadOp` is just a wrapper around the corresponding llvm intrinsic, do we intend to extend those gpu ops for them to do more?

At this case, `MubufLoadOp` can allow i8 type but with minimum `<4xi8>`. I think having `MubufLoadOp` to lower back to `memref::load` would make gpu ops' dependency too low (theoretically speaking gpu ops should not dependent on a "high level op" like from memref.

Reply:
Actually, how about use `miopen.in_bounds_load` here? It was added recently along with `miopen.transforming_for`. The logic here actually looks very identical with what `miopen.in_bounds_load` provides.

Reply:
Looks promising. Let me take a look at it.

Reply:
Swapping to in bound load does make the entire chunk of code much more succinct. We need more of those ops that can make our lives easier.